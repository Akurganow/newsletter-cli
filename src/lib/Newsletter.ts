import * as fm from "front-matter";
import * as fs from "fs";
import * as marked from "marked";

import rmm = require("remove-markdown");

interface IFrontMatterAttributes {
  subject: string;
}

export class Newsletter {
  private name: string;
  private filePath: string;

  constructor(name: string) {
    this.validateName(name);
    this.name = name;
    this.filePath = this.resolveFilePath(name);
  }

  public exists() {
    return fs.existsSync(this.filePath);
  }

  public read(): {
    html: string;
    subject: string;
    text: string;
  } {
    const markdown = fs.readFileSync(this.filePath, "utf8");
    const { body, attributes } = fm<IFrontMatterAttributes>(markdown);
    const html = `<title>${attributes.subject}</title>` + marked(body);
    return {
      html: html.trim(),
      subject: attributes.subject.trim(),
      text: rmm(body).trim(),
    };
  }

  public writeTemplate() {
    fs.writeFileSync(
      this.filePath,
      `---
subject: Define your subject here
---

Write your newsletter in Markdown here
`,
      "utf8",
    );
  }

  public getFilePath() {
    return this.filePath;
  }

  public getName() {
    return this.name;
  }

  private validateName(name: string) {
    if (!name.match("^[a-zA-Z0-9_-]+$")) {
      throw new Error(
        "Newsletter name can include only alphanumerical characters, _ and -",
      );
    }
  }

  private resolveFilePath(name: string): string {
    if (!name.endsWith(".md")) {
      return name + ".md";
    }
    return name;
  }
}
