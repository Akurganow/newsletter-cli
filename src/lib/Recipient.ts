import * as ismail from "isemail";

export class Recipient {
  private email: string;
  constructor(raw: { [key: string]: any }) {
    this.email = raw.email;
    if (!this.email) {
      throw new Error(
        `Recipient ${JSON.stringify(
          raw,
        )} does not contain a column named \`email\``,
      );
    }
    if (!ismail.validate(this.email)) {
      throw new Error(`Recipient ${this.email} is not a valid email address`);
    }
  }
  public getEmail() {
    return this.email;
  }
}
