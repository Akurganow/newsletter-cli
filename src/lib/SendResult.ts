export class SendResult {
  private statuses: IStatus[];
  private recipients: IRecipient[];

  constructor(statuses: IStatus[] = [], recipients: IRecipient[] = []) {
    this.statuses = statuses;
    this.recipients = recipients;
  }

  public count(): number {
    return this.recipients.length;
  }

  public failedCount(): number {
    return this.statuses.filter((s) => s.status !== "Success").length;
  }

  public merge(bulk: SendResult) {
    this.statuses = [...this.statuses, ...bulk.statuses];
    this.recipients = [...this.recipients, ...bulk.recipients];
  }

  public failed(): Array<{ recipient: IRecipient; status: IStatus }> {
    const combined = this.recipients.map((recipient, i) => {
      const status = this.statuses[i];
      if (!status) {
        throw new Error("Internal error: inconsistent state of bulk responses");
      }
      return {
        recipient,
        status,
      };
    });

    return combined.filter((c) => c.status.status !== "Success");
  }
}

interface IRecipient {
  getEmail(): string;
}

interface IStatus {
  error?: string;
  status?: string;
  messageId?: string;
}
