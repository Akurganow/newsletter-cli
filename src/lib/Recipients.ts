import * as csvparse from "csv-parse/lib/sync";
import { Recipient } from "./Recipient";

export class Recipients {
  private data: string;
  private recipients: Recipient[];
  private cursor: number;

  constructor(data: string) {
    this.data = data;
    this.recipients = this.parseRecipients();
    this.cursor = 0;
  }

  public hasNext(): boolean {
    return this.cursor < this.count();
  }

  public next(n: number): Recipient[] {
    const slice = this.recipients.slice(this.cursor, this.cursor + n);
    this.cursor += n;
    return slice;
  }

  public parseRecipients() {
    const rawRecipients = csvparse(this.data, {
      columns: true,
      skip_empty_lines: true,
    });
    return rawRecipients.map((r: any) => new Recipient(r));
  }

  public count() {
    return this.recipients.length;
  }
}
