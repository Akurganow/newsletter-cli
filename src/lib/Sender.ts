import { SendResult } from "./SendResult";

export class Sender {
  private newsletter: INewsletter;
  private transport: ITransport;

  constructor(newsletter: INewsletter, transport: ITransport) {
    this.newsletter = newsletter;
    this.transport = transport;
  }

  public async send(
    source: string,
    recipients: IRecipients,
    progress: ProgressFn,
  ) {
    await this.transport.prepareTemplate({
      ...this.newsletter.read(),
      name: this.getTemplateName(),
    });
    const sendRate = await this.getSendRate();
    const results = new SendResult();
    while (recipients.hasNext()) {
      const batch = recipients.next(sendRate);
      const result = await this.transport.send(
        source,
        batch,
        this.getTemplateName(),
      );
      results.merge(result);
      progress(results.count());
      await sleep(1500);
    }
    return results;
  }

  public async getSendRate(): Promise<number> {
    return await this.transport.getSendRate();
  }

  private getTemplateName() {
    return `newsletterCli-` + this.newsletter.getName();
  }
}

interface IRecipient {
  getEmail(): string;
}

interface IRecipients {
  hasNext(): boolean;
  next(n: number): IRecipient[];
}

interface INewsletter {
  getName(): string;
  read(): {
    html: string;
    text: string;
    subject: string;
  };
}

interface ITransport {
  prepareTemplate(template: {
    name: string;
    html: string;
    text: string;
    subject: string;
  }): Promise<void>;

  getSendRate(): Promise<number>;

  send(
    source: string,
    recipients: IRecipient[],
    templateName: string,
  ): Promise<SendResult>;
}

type ProgressFn = (sent: number) => void;

async function sleep(ms: number) {
  return await new Promise((resolve) => setTimeout(resolve, ms));
}

interface IRecipient {
  getEmail(): string;
}
