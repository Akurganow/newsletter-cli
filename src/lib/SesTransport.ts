import * as aws from "aws-sdk";
import { SendResult } from "./SendResult";

export class SesTransport {
  private ses: aws.SES;

  constructor() {
    this.ses = new aws.SES();
  }

  public async getSendRate(): Promise<number> {
    const quota = await this.ses.getSendQuota().promise();
    if (!quota.MaxSendRate) {
      throw new Error("Cannot read quota.MaxSendRate");
    }
    return quota.MaxSendRate;
  }

  public async prepareTemplate(template: {
    html: string;
    text: string;
    subject: string;
    name: string;
  }): Promise<void> {
    const Template = {
      HtmlPart: template.html,
      SubjectPart: template.subject,
      TemplateName: template.name,
      TextPart: template.text,
    };
    try {
      await this.ses
        .createTemplate({
          Template,
        })
        .promise();
    } catch (err) {
      if (err.code !== "AlreadyExists") {
        throw err;
      }
      await this.ses
        .updateTemplate({
          Template,
        })
        .promise();
    }
  }

  public async send(
    source: string,
    recipients: IRecipient[],
    templateName: string,
  ) {
    try {
      const result = await this.ses
        .sendBulkTemplatedEmail({
          DefaultTemplateData: "{}",
          Destinations: recipients.map((r) => {
            return {
              Destination: {
                BccAddresses: [],
                CcAddresses: [],
                ToAddresses: [r.getEmail()],
              },
              ReplacementTemplateData: "{}",
            };
          }),
          Source: source,
          Template: templateName,
        })
        .promise();

      return new SendResult(
        result.Status.map((item) => {
          return {
            error: item.Error,
            messageId: item.MessageId,
            status: item.Status,
          };
        }),
        recipients,
      );
    } catch (err) {
      return new SendResult(
        recipients.map(() => {
          return {
            error: err,
            status: "SES Error",
          };
        }),
        recipients,
      );
    }
  }
}

interface IRecipient {
  getEmail(): string;
}
