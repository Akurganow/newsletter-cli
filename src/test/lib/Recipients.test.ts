import { Recipient } from "../../lib/Recipient";
import { Recipients } from "../../lib/Recipients";

test("create Recipient with valid email", () => {
  const result = new Recipient({
    email: "a@b.com",
  });
  expect(result.getEmail()).toMatchInlineSnapshot(`"a@b.com"`);
});

test("create Recipient with invalid email", () => {
  const test = () =>
    new Recipient({
      email: "test",
    });

  expect(test).toThrowErrorMatchingInlineSnapshot(
    `"Recipient test is not a valid email address"`,
  );
});

test("create Recipient with missing email", () => {
  const test = () =>
    new Recipient({
      missing: "test",
    });

  expect(test).toThrowErrorMatchingInlineSnapshot(
    `"Recipient {\\"missing\\":\\"test\\"} does not contain a column named \`email\`"`,
  );
});

test("create Recipients from valid CSV", () => {
  expect(new Recipients(`email\na@b.com`).count()).toMatchSnapshot();
  expect(
    new Recipients(`email\na@b.com\n\n\nc@d.com`).count(),
  ).toMatchSnapshot();
  expect(new Recipients(`email,smth\na@b.com,smth`).count()).toMatchSnapshot();
  expect(new Recipients(`email\r\na@b.com`).count()).toMatchSnapshot();
});

test("iterate recipients", () => {
  const r = new Recipients(
    [
      "email",
      ...Array(11)
        .fill(0)
        .map((_, i) => `${i}@test.com`),
    ].join("\n"),
  );
  expect(r.count()).toBe(11);
  expect(r.hasNext()).toBe(true);

  const next5 = r.next(5);
  expect(next5.length).toBe(5);
  expect(next5).toMatchInlineSnapshot(`
Array [
  Recipient {
    "email": "0@test.com",
  },
  Recipient {
    "email": "1@test.com",
  },
  Recipient {
    "email": "2@test.com",
  },
  Recipient {
    "email": "3@test.com",
  },
  Recipient {
    "email": "4@test.com",
  },
]
`);

  expect(r.hasNext()).toBe(true);
  const another5 = r.next(5);
  expect(another5.length).toBe(5);
  expect(another5).toMatchInlineSnapshot(`
Array [
  Recipient {
    "email": "5@test.com",
  },
  Recipient {
    "email": "6@test.com",
  },
  Recipient {
    "email": "7@test.com",
  },
  Recipient {
    "email": "8@test.com",
  },
  Recipient {
    "email": "9@test.com",
  },
]
`);

  expect(r.hasNext()).toBe(true);
  const another1 = r.next(5);
  expect(another1.length).toBe(1);
  expect(another1).toMatchInlineSnapshot(`
Array [
  Recipient {
    "email": "10@test.com",
  },
]
`);

  expect(r.hasNext()).toBe(false);
});
