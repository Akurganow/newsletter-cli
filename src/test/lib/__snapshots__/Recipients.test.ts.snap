// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create Recipients from valid CSV 1`] = `1`;

exports[`create Recipients from valid CSV 2`] = `2`;

exports[`create Recipients from valid CSV 3`] = `1`;

exports[`create Recipients from valid CSV 4`] = `1`;
